components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -799,6 +851,163 @@ void FeedV2Builder::UpdateRequest::AlsoUpdate(
   callbacks.push_back(std::move(callback));
 }
 
+// static
+mojom::FeedV2Ptr FeedV2Builder::GenerateBasicFeed(FeedGenerationInfo info,
+                                                  PickArticles pick_hero,
+                                                  PickArticles pick_article) {
+  DVLOG(1) << __FUNCTION__;
+  DCHECK(!content::BrowserThread::CurrentlyOn(content::BrowserThread::UI));
+
+  auto articles = GetArticleInfos(info.locale, info.feed_items, info.publishers,
+                                  info.signals);
+
+  auto feed = mojom::FeedV2::New();
+
+  constexpr size_t kIterationsPerAd = 2;
+  size_t blocks = 0;
+  while (!articles.empty()) {
+    auto items = GenerateBlock(articles, pick_hero, pick_article,
+                               /*inline_discovery_ratio=*/0);
+    if (items.empty()) {
+      break;
+    }
+
+    // After the first block, every second block should be an ad.
+    if (blocks % kIterationsPerAd == 0 && blocks != 0) {
+      std::ranges::move(GenerateSpecialBlock(info.suggested_publisher_ids),
+                        std::back_inserter(items));
+    }
+
+    std::ranges::move(items, std::back_inserter(feed->items));
+    blocks++;
+  }
+
+  // Insert an ad as the second item.
+  if (feed->items.size() > 1) {
+    auto ad = GenerateAd();
+    feed->items.insert(feed->items.begin() + 1,
+                       std::make_move_iterator(ad.begin()),
+                       std::make_move_iterator(ad.end()));
+  }
+
+  return feed;
+}
+
+// static
+mojom::FeedV2Ptr FeedV2Builder::GenerateAllFeed(FeedGenerationInfo info) {
+  DVLOG(1) << __FUNCTION__;
+  DCHECK(!content::BrowserThread::CurrentlyOn(content::BrowserThread::UI));
```

Comment:
it's not good to add content dependencies to components if they aren't strictly necessary because it makes this code unusable on ios

Reply:
It doesn't work here in a static method, but normally you can get the current sequence in the constructor and use that for sequence checks.